core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -278,9 +278,24 @@ private[deploy] class Master(
       execOption match {
         case Some(exec) => {
           val appInfo = idToApp(appId)
+          val oldState = exec.state
           exec.state = state
-          if (state == ExecutorState.RUNNING) { appInfo.resetRetryCount() }
+
+          if (state == ExecutorState.RUNNING) {
+            assert(oldState == ExecutorState.LOADING,
```

Comment:
Yes -- asserts in the JVM are disabled by default! this would not check anything then. I checked and it's not what the Scala method does though. However the Scala `assert` can be removed at compile time. And it generates an `AssertionError` which is extreme for this situation. 

I think that if you intend it to always be present as argument/state checking, then `require` is good?